netty/src/main/java/io/grpc/transport/netty/BufferingHttp2ConnectionEncoder.java

```diff
@@ -0,0 +1,313 @@
+    return writeHeaders(ctx, streamId, headers, 0, Http2CodecUtil.DEFAULT_PRIORITY_WEIGHT, false,
+                        padding, endStream, promise);
+  }
+
+  @Override
+  public ChannelFuture writeHeaders(ChannelHandlerContext ctx, int streamId, Http2Headers headers,
+                                    int streamDependency, short weight, boolean exclusive,
+                                    int padding, boolean endOfStream, ChannelPromise promise) {
+    if (existingStream(streamId)) {
+      return super.writeHeaders(ctx, streamId, headers, streamDependency, weight, exclusive,
+                                padding, endOfStream, promise);
+    }
+    if (connection().local().acceptingNewStreams()) {
+      assert streamId > largestCreatedStreamId;
+      largestCreatedStreamId = streamId;
+      return super.writeHeaders(ctx, streamId, headers, streamDependency, weight, exclusive,
+                                padding, endOfStream, promise);
+    }
+    if (connection().isGoAway()) {
+      promise.setFailure(
+              connectionError(PROTOCOL_ERROR, "Sending headers after connection going away."));
+      return promise;
+    }
+    PendingStream pendingStream = pendingStreams.get(streamId);
+    if (pendingStream == null) {
+      pendingStream = new PendingStream(ctx, streamId);
+      pendingStreams.put(streamId, pendingStream);
+    }
+    pendingStream.frames.add(new HeadersFrame(headers, streamDependency, weight, exclusive,
+                             padding, endOfStream, promise));
+    return promise;
+  }
+
+  @Override
+  public ChannelFuture writeRstStream(ChannelHandlerContext ctx, int streamId, long errorCode,
+                                      ChannelPromise promise) {
+    if (existingStream(streamId)) {
+      return super.writeRstStream(ctx, streamId, errorCode, promise);
+    }
+    PendingStream stream = pendingStreams.remove(streamId);
+    if (stream != null) {
+      cancelPendingStream(stream);
+      promise.setSuccess();
+    } else {
+      promise.setFailure(connectionError(PROTOCOL_ERROR, "Stream does not exist %d", streamId));
+    }
+    return promise;
+  }
+
+  @Override
+  public ChannelFuture writeData(ChannelHandlerContext ctx, int streamId, ByteBuf data, int padding,
+                                 boolean endOfStream, ChannelPromise promise) {
+    if (existingStream(streamId)) {
+      return super.writeData(ctx, streamId, data, padding, endOfStream, promise);
+    }
+    PendingStream pendingStream = pendingStreams.get(streamId);
+    if (pendingStream != null) {
+      pendingStream.frames.add(new DataFrame(data, padding, endOfStream, promise));
+    } else {
+      data.release();
+      promise.setFailure(connectionError(PROTOCOL_ERROR, "Stream does not exist %d", streamId));
+    }
+    return promise;
+  }
+
+  @Override
+  public ChannelFuture writeGoAway(ChannelHandlerContext ctx, int lastStreamId, long errorCode,
```

Comment:
Do we need to do anything special here or can we just rely on the standard GO_AWAY logic to handle cleaning up the pending streams?